### PR TITLE
Modernize typedefs

### PR DESCRIPTION
**Pull request by marlamb**
The `using` keyword improves readability simply be letting people read from left to right.
Additionally some `typename` keywords have been removed, when they have not been required.

### CODE REVIEW

**Comment by marlamb** [comment]
@rcurtin and @shrit this time I did a full manual go-through, hoping it is a lot less work for you reviewing the PR.

**Comment by shrit** [comment]
@marlamb, I definitely prefer to have `using` instead of typedef, but @rcurtin can decide on this one.

**Comment by rcurtin** [comment]
No strong opinion from my end.  But, `using` does seem to be nicer on the eyes.  Thanks @marlamb!  I will go through and review soon.

**Review by rcurtin — COMMENTED** [review]
Awesome, thanks @marlamb!  I had a couple small comments.  Also, do you think that you can apply these changes to the documentation in `doc/`?  I believe there are a few code snippets that use `typedef` in there, and this could clean them up.

File: src/mlpack/bindings/cli/get_allocated_memory.hpp [hunk 1]
```diff
@@ -44,7 +44,7 @@ void* GetAllocatedMemory(
 {
   // Here we have a model, which is a tuple, and we need the address of the
   // memory.
-  typedef std::tuple<T*, std::string> TupleType;
+  using TupleType =  std::tuple<T*, std::string>;
```
[the remark this answers is not in the document]
Comment (marlamb) [reply src/mlpack/bindings/cli/get_allocated_memory.hpp:47]: Fixed locally.

File: src/mlpack/bindings/cli/delete_allocated_memory.hpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ void DeleteAllocatedMemoryImpl(
     const std::enable_if_t<data::HasSerialize<T>::value>* = 0)
 {
   // Delete the allocated memory (hopefully we actually own it).
-  typedef std::tuple<T*, std::string> TupleType;
+  using TupleType =  std::tuple<T*, std::string>;
```
[the remark this answers is not in the document]
Comment (marlamb) [reply src/mlpack/bindings/cli/delete_allocated_memory.hpp:45]: Fixed locally.

File: src/mlpack/bindings/cli/get_param.hpp [hunk 3]
```diff
@@ -115,7 +115,7 @@ T*& GetParam(
 {
   // If the model is an input model, we have to load it from file.  'value'
   // contains the filename.
-  typedef std::tuple<T*, std::string> TupleType;
+  using TupleType =  std::tuple<T*, std::string>;
```
[the remark this answers is not in the document]
Comment (marlamb) [reply src/mlpack/bindings/cli/get_param.hpp:118]: Found two more and fixed them.

File: src/mlpack/core/math/log_add_impl.hpp [hunk 4]
```diff
@@ -65,7 +65,7 @@ typename T::elem_type AccuLog(const T& x)
   if (maxVal == -std::numeric_limits<typename T::elem_type>::infinity())
     return maxVal;
 
-  return maxVal + std::log(sum(exp(x - maxVal)));;
+  return maxVal + std::log(sum(exp(x - maxVal)));
```
[the remark this answers is not in the document]
Comment (marlamb) [reply src/mlpack/core/math/log_add_impl.hpp:68]: Thanks, just saw it by chance.

File: src/mlpack/core/tree/rectangle_tree/x_tree_auxiliary_information.hpp [hunk 5]
```diff
@@ -165,7 +165,7 @@ class XTreeAuxiliaryInformation
    * The X tree requires that the tree records it's "split history".  To make
    * this easy, we use the following structure.
    */
-  typedef struct SplitHistoryStruct
+  using SplitHistoryStruct = struct SplitHistoryStruct
```
[the remark this answers is not in the document]
Comment (marlamb) [reply src/mlpack/core/tree/rectangle_tree/x_tree_auxiliary_information.hpp:168]: Indeed, I removed it.

File: src/mlpack/bindings/cli/get_param.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/bindings/cli/get_param.hpp]: Awesome, that caught a lot more than I expected.  Thanks!

**Review by rcurtin — APPROVED** [review]
Everything looks great to me.  Thanks again for doing this.  I'll update the `typedef`s in a couple open tree PRs I have been working on (#3814, #3815).

**Review by shrit — APPROVED** [review]
Looks good to me

**Pull request merged**